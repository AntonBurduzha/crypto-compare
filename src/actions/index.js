import { FETCH_CRYPTO_CURRENCIES, GET_NEXT_PAGE } from '../constants';

export const fetchCryptoCurrencies = () => ({ type: FETCH_CRYPTO_CURRENCIES });

export const getNextPage = index => ({ type: GET_NEXT_PAGE, index });
