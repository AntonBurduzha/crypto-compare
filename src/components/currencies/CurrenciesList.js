// @flow
import * as React from 'react';
import { Row, Col, Card } from 'antd';
import { CRYPRO_COMPARE_OLD_API } from '../../constants';
import type { Currency } from '../../types/entities';
import type { Element } from 'react';

type Props = {
  list: Array<Currency>
};

const CurrenciesList = (props: Props): Element<any> =>
  <Row type="flex" justify="start" gutter={16} style={{margin: '0', padding: '0 20px' }}>
    { props.list.map((item: Currency): Element<any> => {
        return (
          <Col span={3} key={item.Id} style={{ margin: '10px 0', padding: '0' }}>
            <Card
              hoverable
              style={{ width: 140, textAlign: 'center' }}
              cover={
                <a href={`${CRYPRO_COMPARE_OLD_API}${item.Url}`} target="_blank">
                  <img alt={item.FullName} src={`${CRYPRO_COMPARE_OLD_API}${item.ImageUrl}`}/>
                </a>
              }
            >
              <Card.Meta title={item.Symbol} description={item.CoinName}/>
            </Card>
          </Col>
        );
      })
    }
  </Row>

export default CurrenciesList;
