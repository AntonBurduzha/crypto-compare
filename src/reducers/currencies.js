// @flow
import { PAGE_SIZE } from '../constants';
import type { Action } from '../types/actions';
import type { CurrenciesState } from '../types/reducers';

export const initState: CurrenciesState = {
  fetching: false,
  error: false,
  fullList: [],
  filteredList: [],
  pageList: [],
  page: 1,
  searchedKey: '',
};

function currenciesReducer(
  state: CurrenciesState = initState,
  action: Action
): CurrenciesState {
  switch (action.type) {
    case 'FETCH_CRYPTO_CURRENCIES': {
      return { ...state, fetching: true, error: false };
    }
    case 'SUCCESS_CRYPTO_CURRENCIES': {
      return {
        ...state,
        fetching: false,
        fullList: action.list,
        pageList: action.list.slice(0, PAGE_SIZE),
      };
    }
    case 'FAILED_CRYPTO_CURRENCIES': {
      return { ...state, fetching: false, error: true };
    }
    case 'GET_NEXT_PAGE': {
      return { ...state, page: action.index };
    }
    case 'SET_SEARCH_CC_VALUE': {
      return { ...state, searchedKey: action.value };
    }
    case 'UPDATE_LIST_BY_SEARCH_VALUE': {
      return {
        ...state,
        pageList: action.pageList,
        filteredList: action.filteredList,
      };
    }
    default:
      return state;
  }
}

export default currenciesReducer;
