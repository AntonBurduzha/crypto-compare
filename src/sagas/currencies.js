import { call, put, takeEvery, select } from 'redux-saga/effects'
import * as constants from '../constants';
import { notification } from 'antd';
import { getCryptoCurrenciesList } from '../api';

function* fetchCurrencies() {
   try {
      const { Data: list } = yield call(getCryptoCurrenciesList);
      yield put({ type: constants.SUCCESS_CRYPTO_CURRENCIES, list: Object.values(list) });
   } catch (e) {
      yield put({ type: constants.FAILED_CRYPTO_CURRENCIES });
      notification.error({
        message: 'Unexpected result!',
        description: 'Something went wrong. Try later or go away from this app.',
      });
   }
}

export function* fetchCurrenciesRequest() {
  yield takeEvery(constants.FETCH_CRYPTO_CURRENCIES, fetchCurrencies);
}

function* setNextPage({ index }) {
  const fullList = yield select(state => state.app.currencies.fullList);
  const isLastPage = Math.ceil(fullList.length / constants.PAGE_SIZE) === index;
  let pageList = [];

  if (index === 1) {
    pageList = fullList.slice(0, constants.PAGE_SIZE);
  } else if (isLastPage) {
    pageList = fullList.slice(((index * constants.PAGE_SIZE) - constants.PAGE_SIZE));
  } else {
    const endIndex = index * constants.PAGE_SIZE;
    const startIndex = endIndex - constants.PAGE_SIZE;
    pageList = fullList.slice(startIndex, endIndex);
  }
  yield put({ type: constants.SET_NEXT_PAGE, pageList, index });
}

export function* getNextPage() {
 yield takeEvery(constants.GET_NEXT_PAGE, setNextPage);
}
