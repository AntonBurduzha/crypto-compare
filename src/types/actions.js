// @flow
import * as types from '../constants';
import type { Currency, CurrencyChartItem, CurrencySocketState } from './entities';

export type FetchCryptoCurrencies = { type: string };

export type SuccessCryptoCurrencies = { type: string, list: Array<Currency> };

export type FailedCryptoCurrencies = { type: string };

export type GetNextPage = { type: string, index: number };

export type SetNextPage = { type: string, pageList: Array<Currency>, index: number };

export type UpdateListByValue = { type: string, filteredList: Array<Currency>, pageList: Array<Currency> };

export type GetNextTab = { type: string, tab: string };

export type GetCurrentCryptoCurrency = { type: string, msg: string };

export type SetCurrentCryptoCurrency = { type: string, state: CurrencySocketState };

export type SetChartData = { type: string, list: Array<CurrencyChartItem> };

export type ResetChartData = { type: string };

export type SearchCryptoCurrencies = { type: string, value: string };

export type Action = (
  FetchCryptoCurrencies |
  SuccessCryptoCurrencies |
  FailedCryptoCurrencies |
  GetNextPage |
  SetNextPage |
  UpdateListByValue |
  GetNextTab |
  GetCurrentCryptoCurrency |
  SetCurrentCryptoCurrency |
  SetChartData |
  ResetChartData |
  SearchCryptoCurrencies
);
