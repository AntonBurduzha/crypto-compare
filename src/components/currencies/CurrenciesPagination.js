import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from'redux';
import * as currenciesActions from '../../actions';
import { PAGE_SIZE } from '../../constants';
import { Pagination } from 'antd';

class CurrenciesPagination extends Component {
  onChangePage = pageNumber => this.props.getNextPage(pageNumber);

  render() {
    const { fullList, page, filteredList, searchedKey } = this.props.currencies;
    const list = searchedKey ? filteredList : fullList;
    return (
      <Fragment>
        { list.length &&
            <Pagination
              total={list.length}
              showTotal={(total, range) => {
                return `${range[0]}-${range[1]} of ${total} items`}
              }
              pageSize={PAGE_SIZE}
              defaultCurrent={1}
              current={page}
              onChange={this.onChangePage}
            />
          }
      </Fragment>
    );
  }
}

function mapStateToProps(state) {
  return { currencies: state.app.currencies };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ ...currenciesActions }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CurrenciesPagination);
