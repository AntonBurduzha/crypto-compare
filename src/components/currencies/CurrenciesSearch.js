// @flow
import * as React from 'react';
import { Input } from 'antd';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as currenciesActions from '../../actions';
import type { Action } from '../../types/actions';

type DispatchProps = { searchCryptoCurrencies: (string) => void };

export class CurrenciesSearch extends React.Component<DispatchProps> {
  onSearch = (value: string): void => this.props.searchCryptoCurrencies(value);

  render() {
    return (
      <Input.Search
        placeholder="input search text"
        enterButton="Search"
        size="large"
        onSearch={this.onSearch}
      />
    );
  }
}

export function mapDispatchToProps(dispatch: Action => void): DispatchProps {
  return bindActionCreators({ ...currenciesActions }, dispatch);
}

export default connect(null, mapDispatchToProps)(CurrenciesSearch);
