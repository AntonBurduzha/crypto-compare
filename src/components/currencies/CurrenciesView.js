import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from'redux';
import { Spin } from 'antd';
import * as currenciesActions from '../../actions';
import { ErrorMessage } from '../shared/ErrorMessage';
import CurrenciesPagination from './CurrenciesPagination';
import CurrenciesList from './CurrenciesList';

class CurrenciesView extends Component {
  componentDidMount() {
    this.props.fetchCryptoCurrencies();
  }

  render() {
    const { fetching, error, pageList } = this.props.currencies;
    return (
      <Spin tip="Loading..." spinning={fetching}>
        <div className="main-container">
          <CurrenciesPagination/>
          <CurrenciesList list={pageList}/>
          { error && <ErrorMessage/> }
        </div>
      </Spin>
    );
  }
}

function mapStateToProps(state) {
  return { currencies: state.app.currencies };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ ...currenciesActions }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CurrenciesView);
