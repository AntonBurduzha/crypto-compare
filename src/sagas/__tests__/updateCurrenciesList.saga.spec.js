import { select, put, takeEvery } from 'redux-saga/effects';
import * as actions from'../../actions';
import { updateCurrenciesList, updateCurrenciesListBySearchValue } from '../../sagas/currencies.saga';

describe('updateCurrenciesList saga', () => {
  const generator = updateCurrenciesList({ value: 'ddd' });
  it('should dispatch the updateListByValue action with updated entity lists', () => {
    const fullList = [
      { CoinName: 's', Name: 'ddd' },
      { CoinName: 'fg', Name: 'ddd1dfds' },
      { CoinName: 'f', Name: 'dd2fsdfew' },
    ];
    const selectDescriptor = generator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const filteredList = [{ CoinName: 's', Name: 'ddd' }, { CoinName: 'fg', Name: 'ddd1dfds' }];
    const pageList = filteredList.slice(0, 32);
    const putDescriptor = generator.next(fullList).value;
    expect(fullList).toBeDefined();
    expect(filteredList).toHaveLength(2);
    expect(fullList).toEqual(expect.arrayContaining(filteredList));
    expect(fullList).toEqual(expect.arrayContaining(pageList));
    expect(filteredList).toEqual(expect.arrayContaining(pageList));
    expect(pageList).toHaveLength(2);

    expect(putDescriptor).toEqual(put(actions.updateListByValue(filteredList, pageList)));
  });
});

describe('updateCurrenciesListBySearchValue saga', () => {
  const generator = updateCurrenciesListBySearchValue();
  it('should start task to watch for SET_SEARCH_CC_VALUE action', () => {
    const takeEveryDescriptor = generator.next().value;
    expect(takeEveryDescriptor).toEqual(takeEvery('SET_SEARCH_CC_VALUE', updateCurrenciesList));
  });
});
