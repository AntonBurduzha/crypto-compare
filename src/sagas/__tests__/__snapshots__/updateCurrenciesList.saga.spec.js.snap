// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateCurrenciesList saga should dispatch the updateListByValue action with updated entity lists 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;
